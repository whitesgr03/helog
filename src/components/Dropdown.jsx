import style from "../styles/Dropdown.module.css";
import Settings from "./Settings";

import button from "../styles/utils/button.module.css";
import image from "../styles/utils/image.module.css";

import PropTypes from "prop-types";
import { Link } from "react-router-dom";

const Dropdown = ({ user }) => {
	const darkTheme = false;
	// const user = useContext(userContext);
	const active = true;
	return (
		<div className={`${darkTheme ? style.dark : ""} ${style.dropdown}`}>
			{user && (
				<div className={style.profile}>
					<div className={style.avatar}>
						{user.name && user.name.charAt(0).toUpperCase()}
					</div>
					{user.name}
				</div>
			)}
			<ul>
				<li>
					<button
						className={`${darkTheme ? button.dark : ""} ${
							button.theme
						}`}
					>
						<span
							className={`${image.icon} ${
								darkTheme ? style.moon : style.sun
							}`}
						/>
						{darkTheme ? "Dark" : "Light"} mode
						<div>
							<div />
						</div>
					</button>
				</li>
				{user && (
					<li>
						<button>
							<span
								className={`${image.icon} ${style.settings}`}
							/>
							Settings
						</button>
					</li>
				)}
				<li>
					{user ? (
						<button>
							<span className={`${image.icon} ${style.logout}`} />
							Logout
						</button>
					) : (
						<Link to="/users/login">
							<span className={`${image.icon} ${style.login}`} />
							Login
						</Link>
					)}
				</li>
			</ul>
			{active && <Settings user={user} />}
		</div>
	);
};

Dropdown.propTypes = {
	user: PropTypes.object,
};

export default Dropdown;
