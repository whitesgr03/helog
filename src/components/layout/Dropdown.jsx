// Packages
import { useState } from 'react';
import PropTypes from 'prop-types';
import { Navigate, Link, useLocation } from 'react-router-dom';

// Styles
import style from '../../styles/layout/Dropdown.module.css';
import button from '../../styles/utils/button.module.css';
import image from '../../styles/utils/image.module.css';

// Components
import Settings from './Settings.jsx';

// Utils
import { handleFetch } from '../../utils/handleFetch.js';

export const Dropdown = ({
	user,
	onUser,
	darkTheme,
	onSwitchColorTheme,
	onCloseDropdown,
	onModel,
	onAlert,
}) => {
	const [loading, setLoading] = useState(null);
	const [error, setError] = useState(null);
	const [activeSettings, setActiveSettings] = useState(false);

	const { pathname: previousPath } = useLocation();

	const handleLogout = async () => {
		setLoading(true);

		const url = `${import.meta.env.VITE_RESOURCE_URL}/account/logout`;

		const options = {
			method: 'POST',
			credentials: 'include',
		};

		const result = await handleFetch(url, options);

		result.success ? onUser(null) : setError(result.message);

		setLoading(false);
		onCloseDropdown();
	};

	const handleActiveSettings = () => setActiveSettings(true);
	const handleCloseSettings = () => setActiveSettings(false);
	return (
		<>
			{error ? (
				<Navigate to="/error" state={{ error, previousPath }} />
			) : (
				<div className={style.dropdown}>
					{user?.name && (
						<div className={style.profile}>
							<div className={style.avatar}>
								{user.name.charAt(0).toUpperCase()}
							</div>
							<span title={user.name}>{user.name}</span>
						</div>
					)}
					<ul>
						<li>
							<button className={button.theme} onClick={onSwitchColorTheme}>
								<span
									className={`${image.icon} ${
										darkTheme ? style.moon : style.sun
									}`}
								/>
								{darkTheme ? 'Dark' : 'Light'} mode
								<div>
									<div />
								</div>
							</button>
						</li>
						{user && (
							<li>
								<button onClick={handleActiveSettings}>
									<span className={`${image.icon} ${style.settings}`} />
									Settings
								</button>
							</li>
						)}
						<li>
							{user ? (
								<button onClick={handleLogout}>
									{loading ? (
										<span
											className={`${image.icon} ${loadingStyles.load} ${loadingStyles['load-desktop']}`}
										/>
									) : (
										<span className={`${image.icon} ${style.logout}`} />
									)}
									Logout
								</button>
							) : (
								<Link to="login" onClick={onCloseDropdown}>
									<span className={`${image.icon} ${style.login}`} />
									Login
								</Link>
							)}
						</li>
					</ul>
					{activeSettings && (
						<Settings
							user={user}
							handleCloseSettings={handleCloseSettings}
						/>
					)}
				</div>
			)}
		</>
	);
};

Dropdown.propTypes = {
	user: PropTypes.object,
	onUser: PropTypes.func,
	darkTheme: PropTypes.bool,
	onSwitchColorTheme: PropTypes.func,
	onCloseDropdown: PropTypes.func,
	onModel: PropTypes.func,
	onAlert: PropTypes.func,
};
