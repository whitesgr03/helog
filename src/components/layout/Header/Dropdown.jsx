// Packages
import { useState } from 'react';
import PropTypes from 'prop-types';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import Cookies from 'js-cookie';
import { useMutation } from '@tanstack/react-query';
import { queryClient } from '../../../utils/queryOptions.js';

// Styles
import styles from './Dropdown.module.css';
import buttonStyles from '../../../styles/button.module.css';
import imageStyles from '../../../styles/image.module.css';
import loadingStyles from '../../utils/Loading.module.css';

// Components
import { Settings } from './Settings.jsx';

// Utils
import { handleFetch } from '../../../utils/handleFetch.js';

// Variables
const URL = `${import.meta.env.VITE_RESOURCE_URL}/account/logout`;

export const Dropdown = ({
	user,
	onAlert,
	darkTheme,
	onColorTheme,
	onCloseDropdown,
	onActiveModal,
}) => {
	const [activeSettings, setActiveSettings] = useState(false);

	const navigate = useNavigate();
	const { pathname: previousPath } = useLocation();

	const { isPending, mutate } = useMutation({
		mutationFn: async () => {
			const options = {
				method: 'POST',
				headers: {
					'X-CSRF-TOKEN': Cookies.get(
						import.meta.env.PROD ? '__Secure-token' : 'token',
					),
				},
				credentials: 'include',
			};
			return await handleFetch(URL, options);
		},
		onError: error =>
			navigate('/error', {
				state: { error: error.message, previousPath },
			}),
		onSuccess: () => queryClient.setQueryData(['userInfo'], null),
		onSettled: () => onCloseDropdown(),
	});

	const handleLogout = () => mutate();

	const handleToggleSettingsMenu = () => setActiveSettings(!activeSettings);

	return (
		<div className={styles.dropdown}>
			{user && (
				<div className={styles.profile}>
					<div className={styles.avatar}>
						{user.username.charAt(0).toUpperCase()}
					</div>
					<span title={user.username}>{user.username}</span>
				</div>
			)}
			<ul>
				<li>
					<button className={buttonStyles.theme} onClick={onColorTheme}>
						<span
							data-testid="theme-icon"
							className={`${imageStyles.icon} ${
								darkTheme ? styles.moon : styles.sun
							}`}
						/>
						{darkTheme ? 'Dark' : 'Light'} mode
						<div>
							<div />
						</div>
					</button>
				</li>
				{user && (
					<li>
						<button onClick={handleToggleSettingsMenu}>
							<span className={`${imageStyles.icon} ${styles.settings}`} />
							Settings
						</button>
					</li>
				)}
				<li>
					{user ? (
						<button onClick={handleLogout}>
							<span
								data-testid="loading-icon"
								className={`${imageStyles.icon} ${isPending ? loadingStyles.load : styles.logout}`}
							/>
							Logout
						</button>
					) : (
						<Link to="../login" onClick={onCloseDropdown}>
							<span className={`${imageStyles.icon} ${styles.login}`} />
							Login
						</Link>
					)}
				</li>
			</ul>
			{activeSettings && (
				<Settings
					user={user}
					onAlert={onAlert}
					onActiveModal={onActiveModal}
					onToggleSettingsMenu={handleToggleSettingsMenu}
					onCloseDropdown={onCloseDropdown}
				/>
			)}
		</div>
	);
};

Dropdown.propTypes = {
	user: PropTypes.object,
	darkTheme: PropTypes.bool,
	onColorTheme: PropTypes.func,
	onCloseDropdown: PropTypes.func,
	onActiveModal: PropTypes.func,
	onAlert: PropTypes.func,
};
