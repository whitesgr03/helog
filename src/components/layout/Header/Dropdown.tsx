// Packages
import { useState } from 'react';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import Cookies from 'js-cookie';
import { useMutation } from '@tanstack/react-query';

// Styles
import styles from './Dropdown.module.css';
import buttonStyles from '../../../styles/button.module.css';
import imageStyles from '../../../styles/image.module.css';
import loadingStyles from '../../utils/Loading.module.css';

// Components
import { Settings } from './Settings.jsx';

// Utils
import { handleFetch } from '../../../utils/handleFetch.js';
import { queryClient } from '../../../utils/queryOptions.js';

import { useAppDataAPI } from '../../pages/App/AppContext.js';

// Variables
const URL = `${import.meta.env.VITE_RESOURCE_URL}/account/logout`;

// Type
import { DarkTheme } from '../../pages/App/App.js';
import { User } from './Header.js';

interface DropdownProps {
	darkTheme: DarkTheme;
	onColorTheme: () => void;
	onCloseDropdown: () => void;
}

export const Dropdown = ({
	darkTheme,
	onColorTheme,
	onCloseDropdown,
}: DropdownProps) => {
	const { onAlert } = useAppDataAPI();
	const [activeSettings, setActiveSettings] = useState(false);

	const { data: user }: { data?: User } =
		queryClient.getQueryData(['userInfo']) ?? {};

	const navigate = useNavigate();
	const { pathname: previousPath } = useLocation();

	const { isPending, mutate } = useMutation({
		mutationFn: async () => {
			const options: RequestInit = {
				method: 'POST',
				headers: {
					'X-CSRF-TOKEN':
						Cookies.get(import.meta.env.PROD ? '__Secure-token' : 'token') ??
						'',
				},
				credentials: 'include',
			};
			return await handleFetch(URL, options);
		},
		onError: () => {
			navigate('/error', { state: { previousPath } });
			onCloseDropdown();
		},
		onSuccess: () => {
			queryClient.removeQueries({ queryKey: ['userInfo'] });
			onAlert([
				{
					message: `You have been logged out.`,
					error: false,
					delay: 4000,
				},
			]);
		},
		onSettled: () => onCloseDropdown(),
	});

	const handleLogout = () => mutate();

	const handleToggleSettingsMenu = () => setActiveSettings(!activeSettings);

	return (
		<div className={styles.dropdown}>
			{user && (
				<div className={styles.profile}>
					<div className={styles.avatar}>
						{user.username.charAt(0).toUpperCase()}
					</div>
					<span title={user.username}>{user.username}</span>
				</div>
			)}
			<ul>
				<li>
					<button className={buttonStyles.theme} onClick={onColorTheme}>
						<span
							data-testid="theme-icon"
							className={`${imageStyles.icon} ${
								darkTheme ? styles.moon : styles.sun
							}`}
						/>
						{darkTheme ? 'Dark' : 'Light'} mode
						<div>
							<div />
						</div>
					</button>
				</li>
				{user && (
					<li>
						<button onClick={handleToggleSettingsMenu}>
							<span className={`${imageStyles.icon} ${styles.settings}`} />
							Settings
						</button>
					</li>
				)}
				<li>
					{user ? (
						<button onClick={handleLogout}>
							<span
								data-testid="loading-icon"
								className={`${imageStyles.icon} ${isPending ? loadingStyles.load : styles.logout}`}
							/>
							Logout
						</button>
					) : (
						<Link to="../login" onClick={onCloseDropdown}>
							<span className={`${imageStyles.icon} ${styles.login}`} />
							Login
						</Link>
					)}
				</li>
			</ul>
			{user && activeSettings && (
				<Settings
					onToggleSettingsMenu={handleToggleSettingsMenu}
					onCloseDropdown={onCloseDropdown}
					user={user}
				/>
			)}
		</div>
	);
};
