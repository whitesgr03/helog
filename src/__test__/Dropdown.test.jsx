import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import Dropdown from "../components/Dropdown";

describe("Dropdown component", () => {
	it("should render the user name, if the 'isLogin' is true", () => {
		const mockUserName = "Jeff";
		const mockIsLogin = true;

		render(<Dropdown isLogin={mockIsLogin} userName={mockUserName} />, {
			wrapper: BrowserRouter,
		});

		const actual = screen.queryByText("Jeff");

		expect(actual).not.toEqual(null);
	});
	it("should render the settings button, if the 'isLogin' is true", () => {
		const mockIsLogin = true;

		render(<Dropdown isLogin={mockIsLogin} />, {
			wrapper: BrowserRouter,
		});

		const actual = screen.queryByRole("button", { name: "Settings" });

		expect(actual).not.toEqual(null);
	});
	it("should render the logout button, if the 'isLogin' is true", () => {
		const mockIsLogin = true;

		render(<Dropdown isLogin={mockIsLogin} />, {
			wrapper: BrowserRouter,
		});

		const actual = screen.queryByRole("button", { name: "Logout" });

		expect(actual).not.toEqual(null);
	});
});
