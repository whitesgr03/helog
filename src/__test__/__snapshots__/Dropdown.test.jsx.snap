// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown component > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="_dropdown_341ed5"
  >
    <ul>
      <li>
        <div
          class="_colorThemeToggle_341ed5"
        >
          <span
            class="icon _sun_341ed5"
          />
          Light mode
          <button
            class="_switchWrap_341ed5"
          >
            <div
              class="_switch_341ed5"
            />
          </button>
        </div>
      </li>
      <li>
        <a
          href="/users/login"
        >
          <span
            class="icon _login_341ed5"
          />
          Login
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
